dali/kernels/shape.h

```diff
@@ -0,0 +1,430 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SHAPE_H_
+#define DALI_KERNELS_SHAPE_H_
+
+#include <array>
+#include <cassert>
+#include <iostream>
+#include <numeric>
+#include <vector>
+
+#include "dali/kernels/span.h"
+
+namespace tensor {
+
+// template <typename T>
+// constexpr typename std::enable_if<std::is_fundamental<T>::value, size_t>::type ShapeDim(const T
+// &) { return 1; }
+
+// template <typename T, size_t N>
+// constexpr int ShapeDim(T (&)[N]) { return int(N); }
+
+// template <typename T>
+// constexpr int ShapeDim(const T &t) {
+//   return int(t.size());
+// }
+
+// /// @brief Calculates flat index of a given element in the tensor
+// /// @remarks If pos has fewer dimensions than shape, the remaining offsets are assumed to be 0
+// template <typename Shape, typename SamplePos>
+// ptrdiff_t CalcOffset(const Shape &shape, const SamplePos &pos) {
+//   ptrdiff_t ofs = pos[0];
+//   const int m = ShapeDim(pos);
+//   const int n = ShapeDim(shape);
+//   int i;
+//   for (i = 1; i < m; i++) {
+//     ofs *= shape[i];
+//     ofs += pos[i];
+//   }
+//   for (; i < n; i++) {
+//     ofs *= shape[i];
+//   }
+//   return ofs;
+// }
+
+/// @brief Returns the product of all elements in shape
+/// @param shape - shape of a tensor whose elements we count
+template <typename T>
+inline int64_t volume(const T &shape) {
+  int n = shape.size();
+  if (n < 1) {
+    return 0;
+  }
+  int64_t v = shape[0];
+  for (int i = 1; i < n; i++) {
+    v *= shape[i];
+  }
+  return v;
+}
+
+constexpr int DynamicDimensions = -1;
+
+template <int ndim>
```

Comment:
How about `int ndim = DynamicDimensions`? This would make the code much shorter. Same applies to Tensor(List)View.
Alternatively, we can come up with a shorter name do `DynamicDimensions` which is exceedingly long.

Reply:
Done